File: tests/tests_py/test_tokenizer.py
import unittest
from forthic.tokenizer import Tokenizer, DLE
from forthic.tokens import StringToken, StartArrayToken, EndArrayToken, StartModuleToken,\
    EndModuleToken, StartDefinitionToken, EndDefinitionToken, WordToken, EOSToken

def get_tokens(tokenizer):
    result = []
    token = None
    while not isinstance(token, EOSToken):
        token = tokenizer.next_token()
        result.append(token)
    return result


def is_word_token(token, name):
    if not isinstance(token, WordToken):
        return False
    if token.name != name:
        return False
    return True


def is_string_token(token, string):
    if not isinstance(token, StringToken):
        return False
    if token.string != string:
        return False
    return True


def is_start_module_token(token, name):
    if not isinstance(token, StartModuleToken):
        return False
    if token.name != name:
        return False
    return True


class TestTokenizer(unittest.TestCase):
    def test_basic(self):
        """Checks to see that all basic tokens are recognized
        """
        tokenizer = Tokenizer("[ ] : DEFINITION ; { } '' WORD")
        tokens = get_tokens(tokenizer)
        expected = [StartArrayToken, EndArrayToken, StartDefinitionToken, EndDefinitionToken,
                    StartModuleToken, EndModuleToken, StringToken, WordToken, EOSToken]

        self.assertEqual(len(expected), len(tokens))

        for i in range(len(tokens)):
            self.assertIsInstance(tokens[i], expected[i])

    def test_end_definition(self):
        """Checks that end definition (;) is recognized even at end of word
        """
        tokenizer = Tokenizer("WORD; WORD2")
        tokens = get_tokens(tokenizer)

        self.assertTrue(is_word_token(tokens[0], "WORD"))
        self.assertIsInstance(tokens[1], EndDefinitionToken)
        self.assertTrue(is_word_token(tokens[2], "WORD2"))


    def test_start_module(self):
        tokenizer = Tokenizer("{ {my-mod")
        tokens = get_tokens(tokenizer)

        self.assertTrue(is_start_module_token(tokens[0], ""))
        self.assertTrue(is_start_module_token(tokens[1], "my-mod"))

    def test_strings(self):

        tokenizer = Tokenizer(f"'Single' ^Caret^ '''Triple Single''' ^^^Triple Caret^^^ {DLE}Single DLE{DLE}")
        tokens = get_tokens(tokenizer)

        self.assertTrue(is_string_token(tokens[0], "Single"))
        self.assertTrue(is_string_token(tokens[1], "Caret"))
        self.assertTrue(is_string_token(tokens[2], "Triple Single"))
        self.assertTrue(is_string_token(tokens[3], "Triple Caret"))
        self.assertTrue(is_string_token(tokens[4], "Single DLE"))

        tokenizer = Tokenizer('"Double" """Triple Double"""')
        tokens = get_tokens(tokenizer)

        self.assertTrue(is_string_token(tokens[0], "Double"))
        self.assertTrue(is_string_token(tokens[1], "Triple Double"))

    def test_arrays(self):
        tokenizer = Tokenizer("[1 2] [3[4]]")

        tokens = get_tokens(tokenizer)

        self.assertTrue(isinstance(tokens[0], StartArrayToken))
        self.assertTrue(is_word_token(tokens[1], "1"))
        self.assertTrue(is_word_token(tokens[2], "2"))
        self.assertTrue(isinstance(tokens[3], EndArrayToken))
        self.assertTrue(isinstance(tokens[4], StartArrayToken))
        self.assertTrue(is_word_token(tokens[5], "3"))
        self.assertTrue(isinstance(tokens[6], StartArrayToken))
        self.assertTrue(is_word_token(tokens[7], "4"))
        self.assertTrue(isinstance(tokens[8], EndArrayToken))
        self.assertTrue(isinstance(tokens[9], EndArrayToken))

    def test_end_module(self):
        tokenizer = Tokenizer("WORD1}WORD2")
        tokens = get_tokens(tokenizer)

        self.assertTrue(is_word_token(tokens[0], "WORD1"))
        self.assertTrue(isinstance(tokens[1], EndModuleToken))
        self.assertTrue(is_word_token(tokens[2], "WORD2"))


if __name__ == '__main__':
    unittest.main()

File: tests/tests_js/test_tokenizer.mjs
import { TOK_STRING, TOK_COMMENT, TOK_START_ARRAY, TOK_END_ARRAY, TOK_START_MODULE, TOK_END_MODULE,
         TOK_START_DEF, TOK_END_DEF, TOK_WORD, TOK_EOS,
         Token, Tokenizer, DLE} from '../../forthic-js/tokenizer.mjs';
import { run_tests } from './utils.mjs';


function get_tokens(tokenizer) {
    let result = [];

    let token = null;
    do {
        token = tokenizer.next_token();
        result.push(token);
    } while (token.type != TOK_EOS)

    return result;
}

function is_word_token(token, string) {
    return (token.type == TOK_WORD && token.string == string);
}

function is_start_mod_token(token, string) {
    return (token.type == TOK_START_MODULE && token.string == string);
}

function is_string_token(token, string) {
    return (token.type == TOK_STRING && token.string == string);
}

// Checks to see that all basic tokens are recognized
function test_basic() {
    let tokenizer = new Tokenizer("[ ] : DEFINITION ; { } '' WORD");
    let tokens = get_tokens(tokenizer);
    let expected = [TOK_START_ARRAY, TOK_END_ARRAY, TOK_START_DEF, TOK_END_DEF,
                    TOK_START_MODULE, TOK_END_MODULE, TOK_STRING, TOK_WORD, TOK_EOS];

    let result = tokens.every((token, index) => token.type == expected[index]);
    return result;
}


// Checks that end definition (;) is recognized even at end of word
function test_end_definition() {
    let tokenizer = new Tokenizer("WORD; WORD2")
    let tokens = get_tokens(tokenizer)

    if (!is_word_token(tokens[0], "WORD"))   return false;
    if (tokens[1].type != TOK_END_DEF)       return false;
    if (!is_word_token(tokens[2], "WORD2"))  return false;

    return true;
}


function test_start_module() {
    let tokenizer = new Tokenizer("{ {my-mod")
    let tokens = get_tokens(tokenizer);

    if (!is_start_mod_token(tokens[0], ""))        return false;
    if (!is_start_mod_token(tokens[1], "my-mod"))  return false;

    return true;
}

function test_strings() {
    let tokenizer = new Tokenizer(`'Single' ^Caret^ '''Triple Single''' ^^^Triple Caret^^^ ${DLE}Single DLE${DLE}`);
    let tokens = get_tokens(tokenizer);
    if (!is_string_token(tokens[0], "Single"))         return false;
    if (!is_string_token(tokens[1], "Caret"))          return false;
    if (!is_string_token(tokens[2], "Triple Single"))  return false;
    if (!is_string_token(tokens[3], "Triple Caret"))   return false;
    if (!is_string_token(tokens[4], "Single DLE"))     return false;

    tokenizer = new Tokenizer('"Double" """Triple Double"""')
    tokens = get_tokens(tokenizer)
    if (!is_string_token(tokens[0], "Double"))         return false;
    if (!is_string_token(tokens[1], "Triple Double"))  return false;

    return true;
}

function test_arrays() {
    let tokenizer = new Tokenizer("[1 2] [3[4]]");
    let tokens = get_tokens(tokenizer);

    if (tokens[0].type != TOK_START_ARRAY)  return false;
    if (!is_word_token(tokens[1], "1"))     return false;
    if (!is_word_token(tokens[2], "2"))     return false;
    if (tokens[3].type != TOK_END_ARRAY)    return false;
    if (tokens[4].type != TOK_START_ARRAY)  return false;
    if (!is_word_token(tokens[5], "3"))     return false;
    if (tokens[6].type != TOK_START_ARRAY)  return false;
    if (!is_word_token(tokens[7], "4"))     return false;
    if (tokens[8].type != TOK_END_ARRAY)    return false;
    if (tokens[9].type != TOK_END_ARRAY)    return false;

    return true;
}

function test_end_module() {
    let tokenizer = new Tokenizer("WORD1}WORD2");
    let tokens = get_tokens(tokenizer);

    if (!is_word_token(tokens[0], "WORD1"))  return false;
    if (tokens[1].type != TOK_END_MODULE)    return false;
    if (!is_word_token(tokens[2], "WORD2"))  return false;

    return true;
}

let tests = {
    "test_basic": test_basic,
    "test_end_definition": test_end_definition,
    "test_start_module": test_start_module,
    "test_strings": test_strings,
    "test_arrays": test_arrays,
    "test_end_module": test_end_module,
}

export { tests };

// run_tests(tests);
